natlib/svcd.c

```diff
@@ -185,3 +192,201 @@ static int svcd_init( lua_State *L )
 
     return 0;
 }
+
+//////////////////////////////////////////////////////////////////////////////
+// SVCD.write((string targetip, number svcid, number attrid, lightfunction payload, 
+//                        number timeout_ms, lightfunction on_done)) 
+// Authors: Aparna Dhinakaran, David Ho, Romi Phadte, Cesar Torres
+/////////////////////////////////////////////////////////////
+
+
+void resolve_table(lua_State *L, char* key){
+    // push the global table, resolve it and then pop it off the stack
+    lua_getglobal(L, "SVCD");
+    lua_pushstring(L, key);
+    lua_gettable(L, -2);
+    lua_insert(L, -2);
+    lua_settop(L, -2);
+}
+
+static int svcd_write_invoke_later(lua_State *L);
+static int svcd_callback_handler(lua_State *L);
+static int anon_function(lua_State *L);
+
+static int svcd_callback_handler(lua_State *L){
+    int ivkd = (int) lua_tonumber(L, lua_upvalueindex(1));
+    printf("making callback function using ivkd = %i \n", ivkd);
+
+    resolve_table(L, "TIMEOUT");
+    int timeout = (int) lua_tonumber(L, -1);
+
+    resolve_table(L, "handlers"); // 7
+    lua_pushnumber(L, ivkd);
+    lua_gettable(L, -2);
+
+    int isnil = (int) lua_toboolean(L, -1);
+    if(isnil != 0){
+        lua_pushnumber(L, timeout);
+        lua_call(L, 1, 0);
+
+        resolve_table(L, "handlers"); // 7
+        lua_pushnumber(L, ivkd);
+        lua_pushnil(L);
+        lua_settable(L, -3);
+    }
+    return 0;
+}
+static int anon_function(lua_State *L){
+    int ivkd = (int)lua_tonumber(L, 1);
+    printf("making anon function using ivkd = %i \n", ivkd);
+
+    lua_pushnumber(L, ivkd);
+    lua_pushcclosure(L, &svcd_callback_handler, 0);
+    return 1;
+}
+
+// write 
+static int svcd_write_invoke_later(lua_State *L){
+    int ivkd = (int)lua_tonumber(L, 1);
+    int timeout_ms = (int)lua_tonumber(L, 2);
+    lua_pushlightfunction(L, libstorm_os_invoke_later);
+    lua_pushnumber(L, timeout_ms * MILLISECOND_TICKS);
+    
+    printf("making custom anon funtion\n");
+    // lua_pushlightfunction(L, anon_function);
+    lua_getglobal(L, "anon_function");
+        lua_pushnumber(L, ivkd);
+        lua_call(L, 0, 1);
+    printf("made custom anon funtion\n");
+
+    lua_call(L, 2, 0);
+    return 0;
+}
+
+static int svcd_write( lua_State *L )
+{   
+
+    printf("Sexy write!\n");    
```

Comment:
This was not on the origin version of the code. Although I like the spirit, this print statement may not be preferable since it might break some people's code if it depends on terminal output. 

Untitled